Libertymakers (#2495)
OpenLiberty/blogs: posts/2022-09-19-app-server-crafters-makers.adoc

---
layout: post

> title: "TITLE"

Reviewer: ```suggestion
title: "Open Liberty- A brief history in maker projects of a cloud-native Java runtime
```

# Do NOT change the categories section
categories: blog
author_picture: https://avatars3.githubusercontent.com/lauracowen
author_github: https://github.com/lauracowen
seo-title: TITLE - OpenLiberty.io

> seo-description: DESCRIPTION

Reviewer: maybe?
```suggestion
seo-description: A look back at some of the interesting and fun projects that crafters and makers on the Liberty team have developed over its 10 year history
```

> blog_description: "DESCRIPTION"

Reviewer: ```suggestion
blog_description: "A look back at some of the interesting and fun projects that crafters and makers on the Liberty team have developed over its 10 year history"
```

open-graph-image: https://openliberty.io/img/blog/crafters-rug-mug-quilt.jpg
---
= Open Liberty: A brief history in maker projects of a cloud-native Java runtime
Laura Cowen <https://github.com/lauracowen>
:imagesdir: /
:url-prefix:
:url-about: /
:figure-caption!:
//Blank line here is necessary before starting the body of the post.

> Five years ago today, in 2017, link:https://openliberty.io/blog/2017/09/19/open-sourcing-liberty.html[IBM published the source code of Liberty], its Java cloud-native runtime, on GitHub as link:https://github.com/openliberty/open-liberty[the Open Liberty project]. Liberty was actually born five years earlier, in 2012 (recently celebrating its 10th birthday), to address the need for a small footprint, extensible, and easily configurable Java runtime that could run on the cloud as well as in a data centre. Liberty was designed to provide link:https://developer.ibm.com/articles/why-cloud-native-java-developers-love-liberty/[a friendly experience to developers] as well as to be powerful and flexible at runtime.

Author: ```suggestion
Five years ago, in 2017, link:https://openliberty.io/blog/2017/09/19/open-sourcing-liberty.html[IBM published the source code of Liberty], its Java cloud-native runtime, on GitHub as link:https://github.com/openliberty/open-liberty[the Open Liberty project]. Liberty was actually born five years earlier, in 2012 (recently celebrating its 10th birthday), to address the need for a small footprint, extensible, and easily configurable Java runtime that could run on the cloud as well as in a data centre. Liberty was designed to provide link:https://developer.ibm.com/articles/why-cloud-native-java-developers-love-liberty/[a friendly experience to developers] as well as to be powerful and flexible at runtime.
```

Author: Please can you also change the publishing date as appropriate and change the slug to `history-maker-projects` before publishing?

At the time, Java runtimes were mostly big and serious, but Liberty was different and Liberty's development team were inspired to make it fun. The crafters and makers in the team got to work on physical projects that used or represented Liberty in creative ways.

== The first Java runtime on a Raspberry Pi

> Liberty was link:https://youtu.be/cBV4tgLgDwg[the first Java runtime to run on a Raspberry Pi]. Even before Liberty's first release, link:https://twitter.com/sjmaple[Simon Maple] (Liberty's first developer advocate) had link:https://youtu.be/cBV4tgLgDwg[run Liberty on a pre-release Raspberry Pi], which had just link:https://en.wikipedia.org/wiki/Raspberry_Pi#Raspberry_Pi[700 Mhz of processing power and 256 MB of memory]. Simon went on to demonstrate Liberty's applicability to the early Internet of Things by running Liberty and MQTT on a Raspberry Pi to link:https://youtu.be/rH_SmgmyVUo[remotely control a smart house 50 miles away]. 

Reviewer: ```suggestion
Liberty was link:https://youtu.be/cBV4tgLgDwg[the first Java runtime to run on a Raspberry Pi]. Even before Liberty's first release, link:https://twitter.com/sjmaple[Simon Maple] (Liberty's first developer advocate) link:https://youtu.be/cBV4tgLgDwg[ran Liberty on a pre-release Raspberry Pi], which had just link:https://en.wikipedia.org/wiki/Raspberry_Pi#Raspberry_Pi[700 Mhz of processing power and 256 MB of memory]. Simon went on to demonstrate Liberty's applicability to the early Internet of Things by running Liberty and link:https://mqtt.org[MQTT] on a Raspberry Pi to link:https://youtu.be/rH_SmgmyVUo[remotely control a smart house 50 miles away]. 
```

[.img_border_light]
.Running Liberty on the Raspberry Pi in 2012. Photo by Simon Maple.
image::img/blog/crafters-libertypi.png[Running Liberty on the Raspberry Pi in 2012,width=50%,align="center"]

Running Liberty on a Raspberry Pi and other hardware with limited resources sparked the imaginations of the makers within the development team.

== Early Liberty conference demos of Java EE on Liberty

> One much-travelled Liberty conference demo was the link:https://github.com/WASdev/skunkworks.libertycar[remote-controlled Liberty Car], the creation of Simon's successor, link:https://twitter.com/tom_will_banks[Tom Banks] (also responsible for link:https://youtu.be/BLkjVzc5u6E[creating a Minecraft user interface to Liberty]). Tom replaced the guts of a remote-controlled car with a Raspberry Pi running Liberty plus a WiFi access point:

Reviewer: ```suggestion
One much-travelled Liberty conference demo was the link:https://github.com/WASdev/skunkworks.libertycar[remote-controlled Liberty Car], the creation of Simon's successor, link:https://twitter.com/tom_will_banks[Tom Banks] (also responsible for link:https://youtu.be/BLkjVzc5u6E[creating a Minecraft user interface to Liberty]). Tom replaced the guts of a remote-controlled car with a Raspberry Pi running Liberty plus a WiFi access point.
```

[.img_border_light]
.Remote controlled car with embedded Arduino running Liberty.
image::img/blog/crafters-liberty-car.jpeg[Remote controlled car with embedded Arduino running Liberty,width=50%,align="center"]

Any conference attendee could connect and control the car from their own smartphone, demonstrating, among other things, the responsiveness of link:https://www.infoq.com/news/2013/06/ee7-websocket-support/[2013's Java EE 7 WebSocket technology]. Java EE, not then known for its small, lightweight footprint, ran easily on the limited hardware resources. Tom's Liberty Car project matured into a conference game where attendees could race multiple Liberty Cars around a large race-track. Tom talks through how it works:

video::LnmjAUEhfX8[youtube]

== Wearable and throwable infrastructures

> Liberty can run with only a tiny footprint, making it friendly to both developer laptops and the cloud. This was enabled by its original OSGi architecture which meant that any app running on Liberty could start with just minimal capabilities and then add and remove features, including features of the recently released Java EE 7, on the fly at run time as needed by the app.

Reviewer: ```suggestion
Liberty can run with only a tiny footprint, making it friendly to both developer laptops and the cloud. This was enabled by its original OSGi architecture, which meant that any app running on Liberty could start with just minimal capabilities and then add and remove features on the fly at run time as needed by the app, including features of the recently released Java EE 7.
```

[.img_border_light]
.Holly presenting in her chef's hat at JFokus. Photo by Kate Stanley.
image::img/blog/crafters-holly-hat-jfokus.jpg[Holly presenting in her chef's hat at JFokus,width=50%,align="center"]

You can find out more from Holly as she talks about her wearable Java runtime:

video::OE5SLt7UlJk[youtube]

After a couple of years of sporting a chef's hat at Enterprise Java conferences, Holly went further with her Java runtime crafting and made link:https://www.infoq.com/presentations/arduino-app-server/[a cuddly throwable Java runtime]. This was an actual cuddly ball with LEDs and a Raspberry Pi embedded in it. She would throw the ball into the audience part-way through her talk and encourage them to chuck it around the room during her presentations.

[.img_border_light]
.Holly about to throw her app server ball whilst presenting at QCon.
image::img/blog/crafters-holly-ball-qcon.JPG[Holly about to throw her app server ball whilst presenting at QCon,width=50%,align="center"]

== Cloud-native Java microservices games

> As the industry moved to microservices, Enterprise Java evolved into the more microservices-friendly link:https://microprofile.io/[Eclipse MicroProfile] and link:https://projects.eclipse.org/projects/ee4j.jakartaee-platform[Eclipse Jakarta EE], the specifications of which the Liberty development team are actively involved in developing and implementing. So Liberty crafters and makers moved to building demos with microservice architectures rather than single-server applications. Open Liberty's new bright space-related branding also lent itself well to 3D-printed spaceships and space-related games (and link:{url-prefix}/blog/2021/09/24/liberty-bikes.html[a virtual bike racing game]).

Reviewer: ```suggestion
As the industry moved to microservices, Enterprise Java evolved into the more microservices-friendly link:https://microprofile.io/[Eclipse MicroProfile] and link:https://projects.eclipse.org/projects/ee4j.jakartaee-platform[Eclipse Jakarta EE], for which the Liberty development team are actively involved in developing and implementing specifications. So Liberty crafters and makers moved to building demos with microservice architectures rather than single-server applications. Open Liberty's new bright space-related branding also lent itself well to 3D-printed spaceships and space-themed games (and link:{url-prefix}/blog/2021/09/24/liberty-bikes.html[a virtual bike racing game]).
```

image::img/blog/crafters-shellcar.jpeg[3D-printed spaceship shell for a toy car,width=50%,align="center"]

link:{url-prefix}/blog/2022/02/16/space-sentry-challenge.html[Space Sentry] was a game built by link:https://github.com/pgunapal[Prashanth Gunapalasingam] and link:https://github.com/fwji[Frank Ji] on Java microservices. Players controlled a laser that was mounted on top of a 3D-printed spaceship to fire at 3D-printed targets. The movements were handled by Lego and Arduinos that communicated over Wi-Fi with three microservices running in Open Liberty containers on a Raspberry Pi.

The three microservices communicated with each other using REST APIs (JAX-RS and MicroProfile Rest Client), and used other link:{url-prefix}/blog/2019/03/28/microprofile22-liberty-19003.html[MicroProfile 2.2] features such as MicroProfile Config, MicroProfile Health, and MicroProfile Metrics to configure and monitor the services and to post results on the leaderboard.

[.img_border_light]
.Space Sentry spaceship and laser. Photo by the Space Sentry team.
image::img/blog/crafters-spacesentry.png[Space Sentry spaceship and laser,width=70%,align="center"]

> The next generation of space-themed microservices-based game came more recently in the form of Space Rover.

Reviewer: ```suggestion
The next generation of space-themed microservices-based games came more recently in the form of Space Rover.
```

[.img_border_light]
.Space Rover game board. Photo by the Space Rover team.
image::img/blog/crafters-spacerover-gameboard-v2.png[Space Rover game board,width=50%,align="center"]

[.img_border_light]
.Space Rover scoreboard.
image::img/blog/crafters-spacerover-scoreboard.jpg[Space Rover scoreboard,width=50%,align="center"]

Space Rover's services make extensive use of web sockets to communicate between the components of the game (Jakarta WebSocket 2.0) and other link:https://jakarta.ee/release/9.1/[Jakarta EE 9.1] technologies including REST endpoints and HTTP endpoints for communicating with the leaderboard and handling the game statistics (Jakarta JAX-RS 3.0), and database injections for interacting with the leaderboard database (Jakarta CDI 3.0).

> link:https://microprofile.io/compatible/5-0/[MicroProfile 5.0] technologies are used to check the status of the services and database (MicroProfile Health 4.0); to configure connections to different services and database (MicroProfile Config 3.0); to retry connection attempts to the database when there are problems (MicroProfile Fault Tolerance 4.0); to record JVM metrics on the Game service and track the latency of the Game service for the health check test (MicroProfile Metrics 4.0); to generate an HTTP client to send game end statistics to the leaderboard (MicroProfile Rest Client 3.0); and to provide REST API documentation and UI for demonstration (MicroProfile OpenAPI 3.0).

Reviewer: ```suggestion
link:https://microprofile.io/compatible/5-0/[MicroProfile 5.0] technologies are used to check the status of the services and database (MicroProfile Health 4.0); to configure connections to different services and database (MicroProfile Config 3.0); to retry connection attempts to the database when there are problems (MicroProfile Fault Tolerance 4.0); to record JVM metrics on the Game service and track the latency of the Game service for the health check test (MicroProfile Metrics 4.0); to generate an HTTP client to send game end statistics to the leaderboard (MicroProfile Rest Client 3.0); and to provide REST API documentation and a UI for demonstration (MicroProfile OpenAPI 3.0).
```
